vmm_core/virt_support_x86emu/src/emulate.rs

```diff
@@ -419,6 +419,18 @@ pub async fn emulate<T: EmulatorSupport>(
                     None,
                 ));
             }
+            err @ x86emu::Error::NonMemoryOrPortInstruction { .. } => {
+                tracelimit::error_ratelimited!(
+                    error = &err as &dyn std::error::Error,
+                    ?instruction_bytes,
+                    physical_address = cpu.support.physical_address(),
+                    "given an instruction that we shouldn't have been asked to emulate - likely a bug in the caller"
+                );
+
+                panic!(
```

Comment:
I don't think a panic is appropriate. I think you should return an error and let the caller fail the VP. We can still see the effective stack as long as the caller is adding the appropriate context.

Reply:
(And the caller does have more context than you have in your panic--e.g., it has the full hypervisor message that initiated the thing. Perhaps virt_mshv is failing to check the flag that indicates the exit is due to something other than the direct instruction execution, for example.)